### ISSUE TITLE
[DX] Form API: Adding `#attributes` to fields of type `container` breaks `#states`

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5157]
This is a sibling issue to #3817...

It is a known issue that `'#markup'` form elements do not support `#states`. So this won't work:

```php
    $form['controller'] = array(
      '#type' => 'checkbox',
      '#title' => t('Show/hide markup element'),
    );

    $form['warning'] = array(
      '#markup' => t('blah blah...'),
      '#states' => array(
        'visible' => array(
          ':input[name="controller"]' => array('checked' => TRUE),
        ),
      ),
    );
```

The workaround is to add a `'#type' => 'container'` around them, and use `#states` on the container form element instead of directly on the markup element.

This works fine:

```php
    $form['controller'] = array(
      '#type' => 'checkbox',
      '#title' => t('Show/hide container'),
    );

    $form['some_warning'] = array(
      '#type' => 'container',
      '#states' => array(
        'visible' => array(
          ':input[name="controller"]' => array('checked' => TRUE),
        ),
      ),
    );
    $form['some_warning']['message'] = array(
      '#markup' => t('blah blah...'),
    );
```

**This issue here is not about solving that problem**...

Everything works as expected with the above-mentioned workaround, until you try adding any class to the container form element (via `#attributes`).

This does NOT work:

```php
    $form['controller'] = array(
      '#type' => 'checkbox',
      '#title' => t('Show/hide container'),
    );

    $form['some_warning'] = array(
      '#type' => 'container',
      '#attributes' => array('class' => array('messages', 'warning')), // Adding this breaks #states.
      '#states' => array(
        'visible' => array(
          ':input[name="controller"]' => array('checked' => TRUE),
        ),
      ),
    );
    $form['some_warning']['message'] = array(
      '#markup' => t('blah blah...'),
    );
```

To make it work again, one needs to manually add the `form-wrapper` class to the container element:

```php
    $form['controller'] = array(
      '#type' => 'checkbox',
      '#title' => t('Show/hide container'),
    );

    $form['some_warning'] = array(
      '#type' => 'container',
      // The "form-wrapper" class is required here, otherwise #states doesn't
      // work.
      '#attributes' => array('class' => array('form-wrapper', 'messages', 'warning')),
      '#states' => array(
        'visible' => array(
          ':input[name="controller"]' => array('checked' => TRUE),
        ),
      ),
    );
    $form['some_warning']['message'] = array(
      '#markup' => t('blah blah...'),
    );
```

**Comment by klonos** [comment]
Simple PR up for review: https://github.com/backdrop/backdrop/pull/3689

**Comment by klonos** [comment]
@laryn mentioned on Zulip:

> @klonos We should make sure the fix doesn't "unfix" the issue that first made that change, I suppose:
https://github.com/backdrop/backdrop-issues/issues/711

**Comment by klonos** [comment]
Thank you @indigoxela for pointing the obvious in the PR 😉 

I've pushed another change, which makes sure that #711 does not regress. To test it, add these form elements to your form:

```php
$form['toggle'] = array(
  '#type' => 'radios',
  '#title' => t('Toggle me'),
  '#options' => array('hide' => 'hide', 'show' => 'show'),
  '#default_value' => 'hide',
);

$form['option'] = array(
  '#type' => 'checkboxes',
  '#title' => t('Options (should not be shown if options hidden)'),
  '#options' => array('one', 'two', 'three'),
  '#attributes' => array('class' => array('some', 'classes')),
  '#description' => t('Some description (should not be shown if options hidden).'),
  '#states' => array(
    'visible' => array(
      ':input[name="toggle"]' => array('value' => 'show'),
    ),
  ),
);
```

Then play with the "Toggle me" radios, and confirm that when the checkboxes are hidden, the respective label and help text are also hidden with them.

**Comment by jenlampton** [comment]
I added the change record for that old issue (and removed it's label) https://docs.backdropcms.org/change-records/functions-theme_radios-and-theme_checkboxes-have-been-removed

**Comment by laryn** [comment]
@klonos The failing tests are the only issue on this one, is that right?

**Comment by klonos** [comment]
@laryn I think so, yes. I'll rebase, in order to get GHA tests, and will work to finalize this.

**Comment by klonos** [comment]
Sorry for taking this long. This is finally done! 🎉

**Comment by laryn** [comment]
I have tested again in the sandbox using the Conditional Fields module as described earlier, and it functions as expected. Also confirmed that all the checks are green! :+1:

**Comment by bradbulger** [comment]
this looks like it fixes #3489 as well, nice. PR looks good to me.

**Comment by quicksketch** [comment]
Thanks @bradbulger and @laryn for the reviews! And @jenlampton for making that change record! I've merged @klonos's https://github.com/backdrop/backdrop/pull/3689 into 1.x and 1.20.x.
